cvmfs/catalog_traversal_parallel.h

```diff
@@ -0,0 +1,457 @@
+    effective_history_depth_ = Parameters::kNoHistory;
+    effective_timestamp_threshold_ = Parameters::kNoTimestampThreshold;
+    bool result = DoTraverse();
+    effective_history_depth_ = this->default_history_depth_;
+    effective_timestamp_threshold_ = this->default_timestamp_threshold_;
+    return result;
+  }
+
+  /**
+   * Starts the traversal process at the catalog pointed to by the given hash
+   * but doesn't traverse into predecessor catalog revisions. This overrides the
+   * TraversalParameter settings provided at construction.
+   *
+   * @param root_catalog_hash  the entry point into the catalog traversal
+   * @return                   true when catalogs were successfully traversed
+   */
+  bool TraverseRevision(const shash::Any &root_catalog_hash,
+                        const TraversalType type = kBreadthFirstTraversal) {
+    effective_history_depth_ = Parameters::kNoHistory;
+    effective_timestamp_threshold_ = Parameters::kNoTimestampThreshold;
+    bool result = Traverse(root_catalog_hash, type);
+    effective_history_depth_ = this->default_history_depth_;
+    effective_timestamp_threshold_ = this->default_timestamp_threshold_;
+    return result;
+  }
+
+  /**
+   * Figures out all named tags in a repository and uses all of them as entry
+   * points into the traversal process.
+   *
+   * @return      true when catalog traversal successfully finished
+   */
+  bool TraverseNamedSnapshots(const TraversalType type = kBreadthFirstTraversal)
+  {
+    UniquePtr<HistoryTN> tag_db;
+    const typename ObjectFetcherT::Failures retval =
+      this->object_fetcher_->FetchHistory(&tag_db);
+    switch (retval) {
+      case ObjectFetcherT::kFailOk:
+        break;
+
+      case ObjectFetcherT::kFailNotFound:
+        LogCvmfs(kLogCatalogTraversal, kLogDebug,
+                 "didn't find a history database to traverse");
+        return true;
+
+      default:
+        LogCvmfs(kLogCatalogTraversal, this->error_sink_,
+                 "failed to download history database (%d - %s)",
+                 retval, Code2Ascii(retval));
+        return false;
+    }
+
+    HashList root_hashes;
+    bool success = tag_db->GetHashes(&root_hashes);
+    assert(success);
+    return TraverseList(root_hashes, type);
+  }
+
+ protected:
+  static uint32_t hasher(const shash::Any &key) {
+    // Don't start with the first bytes, because == is using them as well
+    return (uint32_t) *(reinterpret_cast<const uint32_t *>(key.digest) + 1);
+  }
+
+  bool DoTraverse() {
+    // Optimal number of threads is yet to be determined. The main event loop
+    // contains a spin-lock, so it should not be more than number of cores.
+    threads_process_ = reinterpret_cast<pthread_t *>
+                        (smalloc(sizeof(pthread_t)*num_threads_));
+    for (unsigned int i = 0; i < num_threads_; ++i) {
+      int retval = pthread_create(&threads_process_[i], NULL,
+                                  ProcessQueue, this);
```

Comment:
By convention, other parts of cvmfs call their thread function `Main...`, e.g. `MainProcessQueue`